File: problems/binary_trees/max_path_sum_in_bt/cpp/max_path_sum_in_bt.cpp
#include "max_path_sum_in_bt.h"

#include <algorithm>
#include <climits>
#include <functional>

int maxPathSum(BinaryTree tree) {
  int globalMax = INT_MIN;

  std::function<int(BinaryTree *)> maxPathSumHelper =
      [&](BinaryTree *node) -> int {
    if (!node)
      return 0;

    int leftMax = std::max(maxPathSumHelper(node->left), 0);
    int rightMax = std::max(maxPathSumHelper(node->right), 0);

    int max = node->value + leftMax + rightMax;
    globalMax = std::max(globalMax, max);

    return node->value + std::max(leftMax, rightMax);
  };

  maxPathSumHelper(&tree);

  return globalMax;
}

File: problems/binary_trees/max_path_sum_in_bt/js/max_path_sum_in_bt.js
export const maxPathSum = (() => {
  const maxPathSumHelper = (node, globalMax) => {
    if (!node) return 0;

    const left = Math.max(maxPathSumHelper(node.left, globalMax), 0);
    const right = Math.max(maxPathSumHelper(node.right, globalMax), 0);

    const localMax = left + right + node.value;

    globalMax.value = Math.max(globalMax.value, localMax);

    return node.value + Math.max(left, right);
  };

  return (tree) => {
    const globalMax = { value: -Infinity };
    maxPathSumHelper(tree, globalMax);
    return globalMax.value;
  };
})();
